neqo-transport/src/server.rs

```diff
@@ -620,71 +561,68 @@ impl Server {
             PacketType::OtherVersion => unreachable!(),
             _ => {
                 qtrace!([self], "Not an initial packet");
-                None
+                Output::None
             }
         }
     }
 
     /// Iterate through the pending connections looking for any that might want
     /// to send a datagram.  Stop at the first one that does.
-    fn process_next_output(&mut self, now: Instant) -> Option<Datagram> {
-        qtrace!([self], "No packet to send, look at waiting connections");
-        while let Some(c) = self.waiting.pop_front() {
-            if let Some(d) = self.process_connection(&c, None, now) {
-                return Some(d);
+    fn process_next_output(&mut self, now: Instant) -> Output {
+        let mut callback = None;
+
+        for connection in self.connections.borrow().values() {
+            match connection.borrow_mut().process(None, now) {
+                Output::None => {}
+                d @ Output::Datagram(_) => return d,
+                Output::Callback(next) => match callback {
+                    Some(previous) => callback = Some(min(previous, next)),
+                    None => callback = Some(next),
+                },
             }
         }
 
-        qtrace!([self], "No packet to send still, check wake up times");
-        loop {
-            let connection = self
-                .connections
-                .borrow()
-                .values()
-                .find(|c| c.borrow().needs_waking(now))
-                .cloned()?;
-            let datagram = self.process_connection(&connection, None, now);
-            connection.borrow_mut().woken();
-            if datagram.is_some() {
-                return datagram;
-            }
-        }
+        callback.map_or(Output::None, Output::Callback)
     }
 
     pub fn process(&mut self, dgram: Option<&Datagram>, now: Instant) -> Output {
-        if self.wake_at.map_or(false, |c| c <= now) {
-            self.wake_at = None;
-        }
+        let out = dgram
+            .map_or(Output::None, |d| self.process_input(d, now))
+            .or_else(|| self.process_next_output(now));
+
+        // Clean-up closed connections.
+        self.connections.borrow_mut().retain(|_, c| {
+            if matches!(c.borrow().state(), State::Closed(_)) {
+                // TODO: Is this still needed?
+                c.borrow_mut().set_qlog(NeqoQlog::disabled());
```

Comment:
This would only be necessary if the connection is somehow kept alive through a hanging reference somewhere.  (This being the essential problem with using ref counting.)  All this does is propagate a new qlog reference through the various components, so as long as those are dropped, forcing the reference to be dropped isn't necessary.  You can probably go with `.retain(|_, c| !matches!(c.borrow().state(), State::Closed(_)))`

Reply:
:+1: thank you! 76227d3f